File: src/Jetpac.Core/Keyboard.fs
namespace Jetpac.Core

/// Port of specbolt's keyboard matrix (peripherals/Keyboard.cpp). Only the
/// matrix and its port decode; the SDL keycode table belongs to the shells.
type Keyboard() =
    // Fable does not support 2D arrays; use a jagged array.
    let keys = Array.init 8 (fun _ -> Array.zeroCreate<bool> 5)

    member this.SetKey(row: int, bit: int, pressed: bool) = keys[row][bit] <- pressed

    /// Keyboard responds to any even address.
    member this.In(address: int) : int option =
        if address &&& 1 = 1 then
            None
        else
            let mutable value = 0xFF
            let res = (address >>> 8) &&& 0xFF
            // The low bit in the top address picks the row; with multiple low bits
            // the results AND together.
            for row in 0..7 do
                if res &&& (1 <<< row) = 0 then
                    for bit in 0..4 do
                        if keys[row][bit] then
                            value <- value &&& ~~~(1 <<< bit)

            Some value

File: src/JetpacFR.Web/webroot/Jetpac.Core/Keyboard.js

import { item, fill, initialize } from "../fable_modules/fable-library-js.5.17.0/Array.js";
import { class_type } from "../fable_modules/fable-library-js.5.17.0/Reflection.js";

/**
 * Port of specbolt's keyboard matrix (peripherals/Keyboard.cpp). Only the
 * matrix and its port decode; the SDL keycode table belongs to the shells.
 */
export class Keyboard {
    constructor() {
        this.keys = initialize(8, (_arg) => fill(new Array(5), 0, 5, false));
    }
}

export function Keyboard_$reflection() {
    return class_type("Jetpac.Core.Keyboard", undefined, Keyboard);
}

export function Keyboard_$ctor() {
    return new Keyboard();
}

export function Keyboard__SetKey_289F56A(this$, row, bit, pressed) {
    item(row, this$.keys)[bit] = pressed;
}

/**
 * Keyboard responds to any even address.
 */
export function Keyboard__In_Z524259A4(this$, address) {
    if ((address & 1) === 1) {
        return undefined;
    }
    else {
        let value = 255;
        const res = ((address >> 8) & 255) | 0;
        for (let row = 0; row <= 7; row++) {
            if ((res & (1 << row)) === 0) {
                for (let bit = 0; bit <= 4; bit++) {
                    if (item(bit, item(row, this$.keys))) {
                        value = ((value & ~(1 << bit)) | 0);
                    }
                }
            }
        }
        return value;
    }
}

